File: src/Go/burn_test/samplers.go
package burn_test

import (
	. "../common"
	"log"
	"math"
)

func CreateStandardSizeSampler(rnd *StdRandom) Distribution {
	// Unit of size = 1 byte

	// Small objects: mean = 32B
	dTypical := NewNormalDistribution(rnd, 32, 64)
	// Large objects, the actual size is 2^n, i.e. mean = 2KB
	dLarge := Transform(NewNormalDistribution(rnd, 11, 1), func(x float64) float64 { return math.Pow(2, x) })
	// Extra large objects, the actual size is 2^n, i.e. mean = 64KB
	dxLarge := Transform(NewNormalDistribution(rnd, 16, 1), func(x float64) float64 { return math.Pow(2, x) })

	combined := &CombinedDistribution{
		[]float64{0.99, 0.0099, 0.0001},
		[]Distribution{dTypical, dLarge, dxLarge},
		rnd,
	}
	if err := combined.Check(); err != nil {
		log.Fatal(err)
	}
	return TruncateMin(combined, 0)
}

func CreateStandardTimeSampler(rnd *StdRandom) Distribution {
	// Unit of time = 1 microsecond
	dMethod := ReflectNegatives(NewNormalDistribution(rnd, 0, 0.1))  // ~ Method or a few
	dRequest := ReflectNegatives(NewNormalDistribution(rnd, 0, 1e5)) // ~ Request or a few
	dLongLiving := NewNormalDistribution(rnd, 1e7, 1e7)
	combined := &CombinedDistribution{
		[]float64{0.950, 0.049, 0.001},
		[]Distribution{dMethod, dRequest, dLongLiving},
		rnd,
	}
	if err := combined.Check(); err != nil {
		log.Fatal(err)
	}
	return TruncateMin(combined, 0)
}

File: src/CSharp/Main/BurnTest/Samplers.cs
using System;
using Benchmarking.Common;

namespace GCBurn.BurnTest
{
    public static class Samplers
    {
        public static IDistribution CreateStandardSizeSampler(StdRandom random)
        {
            // Unit of size = 1 byte

            // Small objects: mean = 32B
            var dTypical = new NormalDistribution(random, 32, 64);
            // Large objects, the actual size is 2^n, i.e. mean = 2KB
            var dLarge = new NormalDistribution(random, 11, 1).Transform(x => Math.Pow(2, x)); 
            // Extra large objects, the actual size is 2^n, i.e. mean = 64KB
            var dxLarge = new NormalDistribution(random, 16, 1).Transform(x => Math.Pow(2, x)); 
                    
            var combined = new CombinedDistribution(random, 
                (0.99,   dTypical),
                (0.0099, dLarge),
                (0.0001, dxLarge));
            return combined.TruncateMin(0);
        }

        public static IDistribution CreateStandardTimeSampler(StdRandom random)
        {
            // Unit of time = 1 microsecond, see BurnTester.TimeSamplerFrequency
            var dMethod = new NormalDistribution(random, 0, 0.1).ReflectNegatives(); // ~ Method or a few
            var dRequest = new NormalDistribution(random, 0, 100_000).ReflectNegatives(); // ~ Request or a few
            var dLongLiving = new NormalDistribution(random, 10_000_000, 10_000_000);
            var combined = new CombinedDistribution(random, 
                (0.950, dMethod),
                (0.049, dRequest),
                (0.001, dLongLiving));
            return combined.TruncateMin(0);
        }
    }
}
